fsspec/implementations/http.py

```diff
@@ -142,7 +154,18 @@ async def _ls(self, url, detail=True, **kwargs):
                     # Ignore FTP-like "parent"
                     out.add("/".join([url.rstrip("/"), l.lstrip("/")]))
         if not out and url.endswith("/"):
-            return await self._ls(url.rstrip("/"), detail=True)
+            return await self._ls_real(url.rstrip("/"), detail=True)
+
+        return out
+
+    async def _ls(self, url, detail=True, **kwargs):
+
+        if self.use_listings_cache and url in self.dircache:
```

Comment:
When self.use_listings_cache is False, url will never be in dircache. Perhaps you don't need to store use_listings_cache at all.

Reply:
Indeed, have you seen [`_ls_from_cache`](https://github.com/intake/filesystem_spec/blob/master/fsspec/spec.py#L335) ?

Reply:
Hi Martin. Thanks for spotting this. Indeed, I haven't seen `_ls_from_cache` before and will make some appropriate amendments. I will probably be able to figure it out by looking at other examples how `_ls_from_cache` is used elsewhere.

Reply:
However, I am not sure about the exception raised here?

https://github.com/intake/filesystem_spec/blob/767388a2b87d5281d41d7031618fd733d70d903b/fsspec/spec.py#L351-L353

So, it will trigger on empty directories? Is this intended?

Reply:
Thank you for your response and the explanation. If it will just work (TM), I won't argue.

Reply:
Probably if you try _ls_from_cache with tests on a mock server, you'll find enough edges that, in hindsight, I wouldn't use it here :)

Reply:
Hi Martin,

as you've resolved the conversation, I believe you don't expect me to do anything on this? Please let me know otherwise.

With kind regards,
Andreas.